Test/tutorial/induction-principle.md

```diff
@@ -0,0 +1,779 @@
+`Bind` and `Assign` to allow updating/introducing several variables at once! We do the following
+updates in the AST:
+
+```dafny
+Bind(bvar: string, bval: Expr, body: Expr)   -->   Bind(bvars: seq<string>, bvals: seq<Expr>, body: Expr)
+Assign(avar: string, aval: Expr)             -->   Assign(avars: seq<string>, avals: seq<Expr>)
+```
+
+**Note:** in the [code development](induction-principle/AST.dfy), we actually rename `Expr` to `Expr_Raw` and define
+`Expr` as a subset type of `Expr_Raw` to enforce the fact that in the `Assign` and `Bind`
+cases, we have the same number of variables in the left-hand side as of expressions in
+the righ-hand side.
+
+Updating the interpreter and the induction principle is then
+straightforward. Interestingly, the proofs using the induction principle (`VarUnchanged`,
+`Pure` and `EliminateMulZero`) only require updates to the definitions declared (and not
+defined) by the induction principle: none of the proofs break.
+
+As expected, the `PureNoInductionPrinciple` proof requires the `Assign` and `Bind` cases
+to be updated.
+
+**Note:** if you look at the number of lines of code, you will notice that updating or
+providing the definitions required by the induction principle forces us to modify more lines
+of code that updating the proofs when not using the induction principle. The question then
+is how straightforward it is to update or provide those definitions, compared to the cost
+of updating broken proofs when we don't use the induction principle. So far, we experienced
+more pain when updating the proofs. It is also worth mentionning the fact that, in the
+Dafny-in-Dafny project, proving *any* theorem without the induction principle proved to
+be a big hassle, while the introduction of the induction principle drastically diminished
+the pain.
+
+<a id="pair-of-state"></a>
+## A different kind of proof: manipulating a pair of states
+
+So far, we only studied proofs which require using one context. This changes if we add
+closures. One way of adding support for closures is to introduce a `Value` datatype,
+where closures are defined by the context they capture, a sequence of input variables,
+and their body:
+
+```dafny
+datatype Value
+| Integer(i: int)
+| Closure(ctx: Context, vars: seq<string>, body: Expr)
+```
+
+The problem when verifying compiler transformations is that we can't use Dafny's native
+equality anymore, because local rewritings might modify the body of closures and hence their
+values: we can't reason with a *syntactic* equality. We thus need to introduce a *semantic*
+equality over values. This semantic equality would state that two closures are equal if,
+when called on equal inputs (in the sense of this semantic equality), they evaluate
+to equal results (in the sense of this semantic equality, again), in the spirit of the
+following predicate:
+
+```dafny
+predicate EqValue(cv: Value, cv': Closure)
```

Comment:
```suggestion
predicate EqValue(cv: Value, cv': Value)
  requires cv.Closure? && cv'.Closure?
```
?

Reply:
Thanks for noticing the typo! I renamed the predicate to `EqValue_Closure`, as its purpose is to illustrate the equivalence relation for the closures case (and consistently changed the types to `Closure`)